Initial Android cross-compiler

directhex opened the pull request:
- [x] Only works on Mac due to `--host=` value (we must pass a `--host` triplet when we pass `--target`, and need to infer the correct one to use). This will cause CI failures until fixed, as we only build the runtime on Linux in CI.
- [x] Audit the configure flags, I sorta guessed based on gluing together the AOT flags from iOS/WASM and the JIT flags from Android
- [x] Less brute-force approach to not using the same CFLAGS etc for "real" runtime as cross-compiler. This was a fast way to filter out things I didn't want without breaking iOS etc, but it's not elegant and will cause problems in the future
- [x] Move AOT building to a new subset, so we can do AOT-only builds (i.e. avoid doing full builds on Mac/Windows)
- [ ] Add Windows version (i.e. Windows builds with vcxproj, we need a cross-compiler vcxproj)
- [x] Add Yaml for all the runtime builds (maybe a new yaml, rather than runtime.yml?)

jkoritzinsky committed:
Start creating runtime, apphost, and targeting packs using the new SDK in the NewSFX subset.

jkoritzinsky committed:
Fix generation of runtime package, runtime symbols package, and host symbols package.

jkoritzinsky committed:
Correcly include hosting files and DIA in the shared framework archives and packages

jkoritzinsky committed:
Include PackageOverrides.txt in the ref pack.

jkoritzinsky committed:
Fix product brand prefix.

jkoritzinsky committed:
Add the shared host installer.

jkoritzinsky committed:
Add the hostfxr build and clean up MacOS building.

jkoritzinsky committed:
Rename host and hostfxr projects to ensure we don't collide with the nuget packaging projects.

jkoritzinsky committed:
Add bundle installer.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into adopt-unified-sharedfx

jkoritzinsky committed:
Update platform manifest entries to include linux/mac file variations from the runtime.

jkoritzinsky committed:
Update apphost pack to match master.

jkoritzinsky committed:
Fix typo

jkoritzinsky committed:
Include manpages and fix linux package properties.

jkoritzinsky committed:
Fix linux package properties.

jkoritzinsky committed:
Match short-description from packages in master.

jkoritzinsky committed:
Update to uploaded package.

jkoritzinsky committed:
Update metadata and SDK to generate an exact match of the productbuild xml file.

jkoritzinsky committed:
First draft moving crossgen2 to new SDK.

jkoritzinsky committed:
Get crossgen2 pack working with updated SDK.

jkoritzinsky committed:
Merge remote-tracking branch 'dotnet/master' into adopt-unified-sharedfx

jkoritzinsky committed:
Build archives from the installer and bundle projects.

jkoritzinsky committed:
Move the dotnet-runtime-deps installers to use the new SDK.

jkoritzinsky committed:
Rough attempt to convert the build to use the new sfx sdk.

jkoritzinsky committed:
Remove now empty subsets.

jkoritzinsky committed:
Update installer subsets for testing.

jkoritzinsky committed:
Remove non-existent packaging project from subsets.

jkoritzinsky committed:
Add missing Windows api shim dll to platform manifest.

jkoritzinsky committed:
Update sharedfx sdk.

jkoritzinsky committed:
Add both versions of the long-name dac to the platform manifest.

jkoritzinsky committed:
Skip builds for non-relevant components on Mono.

jkoritzinsky committed:
Remove installer tests by default temporarily to validate product build.

jkoritzinsky committed:
Remove pkgprojs subset from mono installer build since it doesn't build the hosts.

jkoritzinsky committed:
Skip the apphost pack on mono.

jkoritzinsky committed:
Update host test infra to use the new sfx projects.

jkoritzinsky committed:
Delete an out of date test (the alternative case this test exercises is now the only case in this repo)

jkoritzinsky committed:
Fix packaging tests and remove NETStandard packaging tests since we aren't producing that ref pack any more.

jkoritzinsky committed:
Fixes for test official build failures. Re-eneable installer tests by default.

jkoritzinsky committed:
Fix condition for building linux installers.

jkoritzinsky committed:
Exclude Mono header files from the deps.json. They don't need to be in it.

jkoritzinsky committed:
Don't include the host files on the sfxproj.

jkoritzinsky committed:
Add mono-specific files to the platform manifest.

jkoritzinsky committed:
Remove extra copy step in installer job.

jkoritzinsky committed:
Include libmonosgen-2.0.a in platform manifest.

jkoritzinsky committed:
Fix change in linux-x64 test steps.

jkoritzinsky committed:
Fix display name of linux package steps.

jkoritzinsky committed:
Build the installers based on the portable build when applicable so they can find their artifacts.

jkoritzinsky committed:
The installer build no longer consumes Microsoft.NETCore.Platforms and Microsoft.NETCore.Targets, try removing the dependency on the AllConfigurations leg from the installer build.

jkoritzinsky committed:
Set PackageBrandNameSuffix for runtime-deps packages.

jkoritzinsky committed:
Updated sharedfx sdk

jkoritzinsky committed:
Fix debian runtime-deps dependency elements.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into adopt-unified-sharedfx-official-build-test

jkoritzinsky committed:
Add new platform manifest entries.

jkoritzinsky committed:
Add icudt.dat to the platform manifest.

jkoritzinsky committed:
Invert conditions to correctly skip generating installers on non-applicable platforms.

jkoritzinsky committed:
Merge branch 'adopt-unified-sharedfx-official-build-test' of /mnt/d/source/runtime into adopt-unified-sharedfx-official-build-test

jkoritzinsky committed:
Create extra installer subset so we can build just the installers (and avoid double-building nuget packages). Rename the newsfx subset and folder to sfx.

jkoritzinsky committed:
Fix Crossgen2 package name.

jkoritzinsky committed:
Merge branch 'master' into adopt-unified-sharedfx

jkoritzinsky committed:
Clean up some dead MSBuild goo and update the sharedfx package so we can override the crossgen2 package name without breaking the default package id logic.

jkoritzinsky committed:
Update SharedFX SDK.

jkoritzinsky committed:
Merge branch 'master' into adopt-unified-sharedfx

jkoritzinsky committed:
Fix discovery of the built dotnet layout for installer tests. Clean up some MSBuild along this path.

jkoritzinsky committed:
Update to use the new refactored packages.

jkoritzinsky committed:
Fix macOS build failures.

jkoritzinsky committed:
Fix linux installer build.

jkoritzinsky committed:
Upgrade to newest version of the SDK and centrally manage versions.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into adopt-unified-sharedfx-official-build-test

jkoritzinsky committed:
Update SDK to fix build skipping for the shared framework sdk.

jkoritzinsky committed:
Update SDK with full fix for skipbuild with packaging.

jkoritzinsky committed:
Another try for fixing the build-skip infra.

jkoritzinsky committed:
Upgrade SDK.

jkoritzinsky committed:
Fix package ids for mono packages.

jkoritzinsky committed:
Fix installer tests after update.

jkoritzinsky committed:
Add System.Private.Runtime.InteropServices.Javascript.dll to platform manifest

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into adopt-unified-sharedfx

jkoritzinsky committed:
Fix Crossgen2 package to place all files in the tools folder.

jkoritzinsky committed:
Update to SDK with validation targets.

jkoritzinsky committed:
Add new ICU files to manifest.

jkoritzinsky committed:
Crossgen2 doesn't need to go through framework resolution so we can permit non-file-versioned dlls. (jitinterface.dll is currently unversioned).

Signed-off-by: Jeremy Koritzinsky <jekoritz@microsoft.com>

jkoritzinsky committed:
Put the property in a property group.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into adopt-unified-sharedfx

jkoritzinsky committed:
Ensure we build corehost before sfx since we no longer directly build the project from the sfx projects.

jkoritzinsky committed:
Add missing entry for additional ICU data file.

jkoritzinsky committed:
Remove step downloading all artifacts and remove MSBuild glue to hook up looking up artifacts in the full download.

Signed-off-by: Jeremy Koritzinsky <jekoritz@microsoft.com>

jkoritzinsky committed:
Merge branch 'master' into adopt-unified-sharedfx

jkoritzinsky committed:
Update SDK

Signed-off-by: Jeremy Koritzinsky <jekoritz@microsoft.com>

jkoritzinsky committed:
Update SDK to get MSI fix.

Signed-off-by: Jeremy Koritzinsky <jekoritz@microsoft.com>

jkoritzinsky committed:
Merge branch 'master' into adopt-unified-sharedfx

jkoritzinsky committed:
Don't even try building the hosting, crossgen2, or bundle installers when RuntimeFlavor is Mono.

jkoritzinsky committed:
Remove out of date comment.

Signed-off-by: Jeremy Koritzinsky <jekoritz@microsoft.com>

vargaz commented on src/mono/mono.proj:
```diff
@@ -750,7 +772,19 @@
-      <_MonoAotCrossConfigureCommand>$(MonoProjectRoot)configure @(_MonoAotCrossConfigureParams, ' ') @(_MonoAotCrossAC_VARS, ' ') $(_MonoAotCrossCFLAGSOption) $(_MonoAotCrossCXXFLAGSOption) $(_MonoAotCrossCPPFLAGSOption) $(_MonoAotCrossCXXCPPFLAGSOption) $(_MonoAotCrossLDFLAGSOption) $(_MonoAotCrossCCLDFLAGSOption) $(_MonoCCOption) $(_MonoCXXOption) $(_MonoAROption) $(_MonoASOption) $(_MonoCPPOp ...
+      <_MonoAotCrossCCOption Condition="'$(TargetsAndroid)' != 'true'">$(_MonoCCOption)</_MonoAotCrossCCOption>
+      <_MonoAotCrossCXXOption Condition="'$(TargetsAndroid)' != 'true'">$(_MonoCXXOption)</_MonoAotCrossCXXOption>
+      <_MonoAotCrossAROption Condition="'$(TargetsAndroid)' != 'true'">$(_MonoAROption)</_MonoAotCrossAROption>
```
We don't have most of these options in the cmake build.

vargaz commented on src/mono/netcore/sample/Android/Makefile:
```diff
@@ -12,7 +12,9 @@ run:
 	/p:TargetOS=Android \
 	/p:Configuration=$(MONO_CONFIG) \
-	/p:DeployAndRun=true
+	/p:DeployAndRun=true \
```
The samples should not use aot imho.

vargaz commented:
Comments:
- the -O=-float32 argument to the aot compiler should be removed
- -O=-simd should be added to work around a ton of SIMD problems
- it would be nice to print out the assembly paths for the assemblies which fail aot, its hard to find out which ones because of the parallelism.

steveisok commented:
@directhex Let's try to land the android x64 host + all android targets in a separate PR.  If OSX comes along cheaply, I'm fine with including that as well.  The rest I'll list in an issue and we can chip away w/ individual PR's.

jkoritzinsky committed:
Merge branch 'master' of github.com:dotnet/runtime into adopt-unified-sharedfx

jkoritzinsky committed:
Rename subsets based on PR discussion.

jkoritzinsky committed:
Update platform manifest and fix path to 3rd party notices for installer.

jkoritzinsky committed:
Remove test feed.

jkoritzinsky committed:
Update subset references I missed.

jkoritzinsky committed:
Merge branch 'master' of github.com:dotnet/runtime into adopt-unified-sharedfx

jkoritzinsky committed:
Fixed the subset arguments in other yaml files.

jkoritzinsky committed:
Merge branch 'master' of github.com:dotnet/runtime into adopt-unified-sharedfx

jkoritzinsky committed:
Merge branch 'master' of github.com:dotnet/runtime into adopt-unified-sharedfx

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross2

directhex committed:
Redesign mono.aotcross building

directhex committed:
Rejiggle so AOT compiler gets copied to HOST/cross/TARGET dir

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross2

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross2

directhex committed:
Build a package with the cross compiler inside (hardcoded to android-x64 thus far)

directhex committed:
Parallelize, apparently successfully

directhex committed:
Pull in changes from the other PR

directhex committed:
Only build cross-compiler packs on Linux FOR NOW, until we have NDK on Win/Mac

directhex committed:
Fix building locally

directhex committed:
Fix up subset behaviour

directhex committed:
Build cross in pipeline, and enable cross packages on Mac

directhex committed:
Nope, build cross on osx_x64 job, not android_x64 job

directhex commented:
I expect this to fail right now on not-linux_x64. Need to figure out how to set android_ndk_root for mac (it's in our images already but not exported as an env var)

directhex committed:
Release not debug?

directhex committed:
Fix old-style Android build

directhex committed:
Workaround for x86 builds

directhex committed:
Try something else

directhex committed:
Mumble mumble

directhex committed:
Gosh darn it

directhex committed:
Rename android offset file to help disambiguate

directhex committed:
Create a new property which can generate offsets only

directhex committed:
Revert "Nope, build cross on osx_x64 job, not android_x64 job"

This reverts commit 2364432941eac0ffa9f2193868205c59cded79d7.

directhex committed:
Revert "Nope, build cross on osx_x64 job, not android_x64 job"

This reverts commit 2364432941eac0ffa9f2193868205c59cded79d7.

directhex committed:
Clean up more pipeline errors

directhex committed:
Add new yaml branch for generating offset files

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross

directhex committed:
Fix sysroot setting

vargaz commented on src/mono/mono.proj:
```diff
@@ -457,7 +471,7 @@
       <_MonoRuntimeArtifacts Include="$(_MonoAotCrossFilePath)">
-        <Destination>$(RuntimeBinDir)cross\mono-aot-cross</Destination>
+        <Destination>$(RuntimeBinDir)cross\$(TargetOS)-$(TargetArchitecture)\mono-aot-cross</Destination>
       </_MonoRuntimeArtifacts>
```
Why is this needed ? Isn't the runtime pack already os/arch specific ?

directhex replied:
We need a clean way to have multiple cross compilers next to each other, in order to enable a user to build an apk which supports multiple target architectures (which is common on Android).

vargaz commented on src/mono/mono.proj:
```diff
@@ -466,10 +480,10 @@
       <_MonoRuntimeArtifacts Condition="'$(MonoAOTBundleLLVMOptimizer)' == 'true'" Include="$(MonoAOTLLVMDir)\bin\llc">
-        <Destination>$(RuntimeBinDir)cross\llc</Destination>
+        <Destination>$(RuntimeBinDir)cross\$(TargetOS)-$(TargetArchitecture)\llc</Destination>
       </_MonoRuntimeArtifacts>
```
llc/opt are not target arch specific.

directhex replied:
Assuming that this isn't true is easier for now, i.e. put "the `llc/opt` needed for this `mono-aot-cross` in the same bundle" rather than dependency tracking & public artifact publishing of the llvm binaries for end user consumption. We can clean that story up later.

vargaz commented on src/mono/monoaotcross.proj:
```diff
@@ -0,0 +1,40 @@
+<Project Sdk="Microsoft.Build.Traversal" DefaultTargets="BuildMonoCrossAllTargets">
```
Why is this needed ? Why can't the cross compilers be built normally as on ios ?

directhex replied:
iOS assumes a 1:1 mapping between "host OS/arch" and "target OS/arch". "Build the iOS ARM64 cross compiler" _always_ means "Build the macOS x64 to iOS ARM64 cross compiler". The build pipelines and directory structures are built around this 1:1 assumption.

Right now, this would mean we can only build _one_ Android x64 cross compiler - Linux x64 (the build platform) to Android x64.

We need to decouple the entire 1:1 assumption in order to be able to build multiple hosts for the same target.

directhex commented:
/azp run runtime

directhex committed:
Missing escaping

directhex committed:
Why can't PublishPipelineArtifact do wildcards>?!

directhex committed:
Fic typos

directhex committed:
See if dependencies work

directhex committed:
Ensure Mono jobs pass through dependsOn

directhex committed:
Try adding AotCross to build. This should fail as we don't consume offsets yet

directhex committed:
Try downloading offsets

directhex committed:
Don;t hide behind conditional

directhex committed:
Fix offset aertifact name

directhex committed:
Download to correct place

directhex committed:
Copy offset file where it needs to be

directhex committed:
debug

directhex committed:
add debug, fix path

directhex committed:
typo

directhex committed:
Fix escaping

directhex committed:
Fix building cross compilers with no NDK, but an offset file

directhex committed:
Get rid of debug noise

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross

directhex committed:
Fix noise on Windows

directhex committed:
Build cross on correct platforms

directhex committed:
Build all android targets

directhex committed:
Try and work around llvm-config not being found on windows

directhex committed:
Revert "Get rid of debug noise"

This reverts commit 2f9a661e7d214a2fd6e91a5eefc89f2adec43097.

directhex committed:
More debug spew, osx should be working

directhex committed:
tpyo

directhex committed:
tweaky

directhex committed:
Get rid of a debug output

directhex commented:
Working:
- Linux x64
- OSX x64

Failing:
- Windows x64
- Windows x86

Building, but building wrong:
- Linux arm64

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross

directhex committed:
Shelve fixing Windows and non-x64 hosts for now, get what works fully working

directhex commented:
Punting everything except Mac x64 and Linux x64 host platform builds, to speed things along.

directhex committed:
Remove debug spew

directhex committed:
Make LLVM AOT compatible with new path structure?

akoeplinger committed:
Merge branch 'master' into android-cross

# Conflicts:
#	eng/pipelines/mono/templates/build-job.yml
#	src/mono/mono.proj
#	src/mono/netcore/sample/Android/AndroidSampleApp.csproj

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into android-cross

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into android-cross

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into android-cross

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into android-cross

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into android-cross

akoeplinger committed:
Simplify logic

akoeplinger committed:
Fix Android sample app

akoeplinger committed:
Fix including the mono-cross-aot in the live-built runtime pack

Also put llvm binaries into a non-RID folder since llvm can target multiple platforms.

akoeplinger committed:
Revert changes to MonoAOTCompiler.props

akoeplinger committed:
Revert unnecessary change

akoeplinger committed:
Simplify and fix .yml

akoeplinger committed:
Simplify and fix .yml

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into android-cross

directhex committed:
If we only build AOT compilers in Release, generate Release offsets

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross

directhex committed:
Debug?

directhex committed:
This is a horrible workaround

steveisok commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj:
```diff
@@ -0,0 +1,54 @@
+    <SkipValidatePackage>true</SkipValidatePackage>
+    <PlatformPackageType>RuntimePack</PlatformPackageType>
+    <SharedFrameworkName>Microsoft.NETCore.App.MonoCrossAOT</SharedFrameworkName>
+    <OverridePackageId>Microsoft.NETCore.App.MonoCrossAOT.$(RuntimeIdentifier).targeting.$(TargetCrossRid)</OverridePackageId>
```
This should be `Microsoft.NetCore.App.Runtime.AOT.{host-rid}.Cross.{target-rid}.{version}`

directhex committed:
Rename package

directhex committed:
Remove obsolete debug spew

directhex committed:
New debug spew - this works fine locally, so why broken on azdo?

directhex committed:
More debug

directhex committed:
Stop setting an OS override to Android, it breaks zipping

steveisok reviewed (approved):
Looks good to me

ViktorHofer commented:
181 commits 🤨?

directhex commented:
It was originally based on a long term feature branch not master, that can confuse the GH UI 

Sent from my iPhone

> On Jan 6, 2021, at 9:02 AM, Viktor Hofer <notifications@github.com> wrote:
> 
> ﻿
> 181 commits 🤨?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jkoritzinsky commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj:
```diff
@@ -0,0 +1,54 @@
+    <Error Condition="'$(TargetCrossRid)' == ''" Text="TargetCrossRid not set" />
+  </Target>
+
+  <Target Name="GetFilesToPublish">
```
What was the need for adding this target? At first glance it looks like it matches the one in the SDK.

jkoritzinsky commented on src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.Build.Traversal" DefaultTargets="BuildMonoCrossAllTargets">
```
If we’re using the traversal SDK, then we should hook into the Traversal targets instead of defining our own.

directhex committed:
Clean up some packaging errors

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross

directhex commented:
@jkoritzinsky not really using the traversal SDK, just a leftover from a prior incarnation of that XML, which didn't _break_ whilst not being correct either.

jkoritzinsky commented on src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:
```diff
@@ -0,0 +1,29 @@
+<Project Sdk="Microsoft.NET.Sdk" DefaultTargets="BuildMonoCrossAllTargets">
+  <Sdk Name="Microsoft.DotNet.SharedFramework.Sdk" />
```
If you're not going to use any of the built-in targets, then I'd suggest just using the NoTargets SDK.

```suggestion
<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="BuildMonoCrossAllTargets">
```

jkoritzinsky commented on src/mono/monoaotcross.proj:
```diff
@@ -0,0 +1,42 @@
+<Project Sdk="Microsoft.Build.Traversal" DefaultTargets="BuildMonoCrossAllTargets">
```
Same here with the SDK.

```suggestion
<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="BuildMonoCrossAllTargets">
```

jkoritzinsky commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj:
```diff
@@ -0,0 +1,40 @@
+    <NativeRuntimeAsset Include="$(MonoCrossAOTDir)$(TargetCrossRid)\**" TargetPath="tools/" />
+  </ItemGroup>
+
+  <Target Name="AddRuntimeFilesToPackage" DependsOnTargets="ResolveRuntimeFilesFromLocalBuild">
+  </Target>
+
+  <Target Name="AddFrameworkFilesToPackage" DependsOnTargets="ResolveLibrariesFromLocalBuild">
+  </Target>
```
If you aren't putting anything in these targets then you can remove them and remove setting `GetSharedFrameworkFilesForReadyToRunDependsOn`.

In addition, if all the files are native, then there are more optimizations we can make here to reduce the work we do.

jkoritzinsky commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj:
```diff
@@ -0,0 +1,40 @@
+  <Target Name="AddFrameworkFilesToPackage" DependsOnTargets="ResolveLibrariesFromLocalBuild">
+  </Target>
+
+  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
+  <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.SharedFramework.Sdk" />
+
+  <Target Name="ValidateProperties" BeforeTargets="GenerateNuspec">
+    <Error Condition="'$(TargetCrossRid)' == ''" Text="TargetCrossRid not set" />
+  </Target>
```
You should be able to put this before the `<Import>` tags above and switch the `<Import>` tags back to the `Sdk="Microsoft.NET.Sdk"` attribute and the `<Sdk Name="Microsoft.Dotnet.SharedFramework.Sdk" />` tag.

directhex replied:
Can you check the fixes for this one (e157f3f72c5) are correct?

jkoritzinsky commented:
Just a few cleanup suggestions, but other than those LGTM.

directhex committed:
Fix AOT on iOS

directhex committed:
Cleanup suggestions from Jeremy

directhex committed:
More cleanup

directhex committed:
Fix for WASM failure

directhex committed:
Add Mono cross to official build

directhex committed:
jobParameters cannot be empty

directhex committed:
Be more aggressive in making offsets build

directhex committed:
OK, maybe now?

directhex committed:
Switch monoaotcross.proj back to Traversal, due to NuGet error

directhex commented:
@jkoritzinsky I've switched monoaotcross.proj back to Traversal. With NoTargets, the following error happens during a clean-tree build:

```
                     /Users/directhex/Projects/runtime/.dotnet/sdk/5.0.100/NuGet.RestoreEx.targets(19,5): error : Invalid framework identifier ''. [/Users/directhex/Projects/runtime/Build.proj]
```

jkoritzinsky commented:
I'm that case I'd suggest hooking the defined targets in the project into the targets in the Traversal SDK, but that's not required.

ViktorHofer commented:
> With NoTargets, the following error happens during a clean-tree build:

The NoTargets SDK needs a TargetFramework specified. If you want to stick with NoTargets you can fix it by adding this code block:
```
<PropertyGroup>
    <TargetFramework>$(NetCoreAppToolCurrent)</TargetFramework>
</PropertyGroup>
```

directhex committed:
Better fix from Viktor

directhex commented:
Thanks @ViktorHofer

directhex committed:
Ensure AOT arg is passed to package build

directhex committed:
Did I miss build args?

steveisok commented:
The windows test failure will be resolved by https://github.com/dotnet/runtime/pull/46828

directhex committed:
fixed?

directhex committed:
Change runtime.yml for new checkout behaviour

directhex committed:
Let's not put debug spew into the repo, eh?

directhex committed:
Maybe specifying nameSuffix will allow global-build-job to disambiguate

directhex committed:
Liberal sprinkle of ToLower()

directhex committed:
Merge branch 'android-cross' of github.com:directhex/runtime into android-cross

directhex committed:
Missing args

directhex committed:
Add dependsOn to global-build-job

directhex committed:
Revert "Add dependsOn to global-build-job"

This reverts commit 15cf3c0e7a9c16b0418ec2939c3c3012cf7ac9f7.

directhex committed:
Use build-job directly, going via global-build-job is too hard

safern commented on eng/pipelines/mono/templates/build-job.yml:
```diff
@@ -13,6 +13,9 @@ parameters:
   isOfficialBuild: false
   crossBuild: false
   crossrootfsDir: ''
+  dependOnEvaluatePaths: false
```
This is outdated, I updated this to: `dependOnCheckoutBundle` here: https://github.com/dotnet/runtime/commit/604660b37a0d416155a97cbd09c1edc101a1a1cb

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:
```diff
@@ -0,0 +1,28 @@
+    <MonoAotTargetRids Include="$(MonoAotTargets.Split(';'))" />
+  </ItemGroup>
+
+  <Target Name="RestoreSubProject">
+    <MSBuild Targets="Restore"
+             Projects="$(MSBuildThisFileDirectory)Microsoft.NETCore.App.MonoCrossAOT.sfxproj" />
+  </Target>
```
This causes a separate restore invocation to be triggered which we don't want for different reasons (reliability, performance, etc.). Why is this necessary? If you use the Traversal SDK, you can just add ProjectReference to that project if you don't want to expose it in Subsets.props.

directhex replied:
The build was failing without the explicit second call to restore. I know it's ugly, but it's the workaround I found which worked.

Your suggestion is adding a ProjectReference to Microsoft.NETCore.App.MonoCrossAOT.sfxproj in monocrossaot.sfxproj, and switching Microsoft.NETCore.App.MonoCrossAOT.sfxproj to Traversal instead of NoTargets?

ViktorHofer replied:
> Your suggestion is adding a ProjectReference to Microsoft.NETCore.App.MonoCrossAOT.sfxproj in monocrossaot.sfxproj, and switching Microsoft.NETCore.App.MonoCrossAOT.sfxproj to Traversal instead of NoTargets?

Right, but only if that project isn't restored yet. What was the failure that you were hitting?

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/monocrossaot.sfxproj:
```diff
@@ -0,0 +1,28 @@
+<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="BuildMonoCrossAllTargets">
```
You don't need a default target here and just hook that target as `AfterTargets="Build"` to run after the build which is the default target.

ViktorHofer replied:
@jkoritzinsky just to make sure, the sfxproj SDK doesn't change the default target, right?

jkoritzinsky replied:
You are correct. The SharedFX sdk still uses the Build target as the entry point.

ViktorHofer commented on src/mono/monoaotcross.proj:
```diff
@@ -0,0 +1,46 @@
+<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="BuildMonoCrossAllTargets">
```
Same comment about the default target. Just remove the attribute here and use an aftertargets="Build" in the `BuildMonoCrossAllTargets` target.

vargaz commented on src/mono/netcore/sample/Android/Makefile:
```diff
@@ -1,6 +1,8 @@
 DOTNET := ../../../../.././dotnet.sh
+USE_LLVM=true
+FORCE_AOT=false
 
```
Other samples name this variable 'AOT'.

vargaz commented on src/mono/netcore/sample/iOS/Makefile:
```diff
@@ -1,7 +1,8 @@
-USE_LLVM=True
+USE_LLVM=true
+FORCE_AOT=false
 
```
Ditto.

directhex committed:
Revert "Use build-job directly, going via global-build-job is too hard"

This reverts commit 0587cba933a4268f4de8f6f089a13bd9a2ef5430.

directhex committed:
Move artifact download to global-job, because reasons

directhex committed:
Pass through dependsOn in global-build-job

directhex commented on src/mono/mono.proj:
```diff
@@ -417,18 +431,21 @@
       <_MonoAotCMakeBuildCommand Condition="'$(MonoVerboseBuild)' == 'true'">$(_MonoAotCMakeBuildCommand) --verbose</_MonoAotCMakeBuildCommand>
       <_MonoAotCMakeBuildCommand Condition="'$(_MonoUseNinja)' != 'true'">$(_MonoAotCMakeBuildCommand) --parallel $([System.Environment]::ProcessorCount)</_MonoAotCMakeBuildCommand>
+      <_MonoAotPrebuiltOffsetsFile>$(ArtifactsObjDir)\monooffsetfiles\$(PlatformConfigPathPart)\cross\$([System.IO.Path]::GetFileName('$(MonoAotOffsetsFile)'))</_MonoAotPrebuiltOffsetsFile>
     </PropertyGroup>
```
It can be. Originally it was a little easier for local testing workflow if rm -r artifacts/bin/mono/ didn't touch my local copy of the offset files

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross

directhex committed:
Changes from Zoltan

directhex committed:
Changes from Zoltan

directhex committed:
changes from santi

directhex committed:
Don't redefine

directhex committed:
Better dependsOn

directhex committed:
Typo

directhex committed:
More mess

directhex committed:
Merge remote-tracking branch 'origin/master' into android-cross

directhex committed:
Changes from Viktor

directhex commented:
Green with correct nupkgs on internal https://dev.azure.com/dnceng/internal/_build/results?buildId=950402&view=results

directhex commented:
None of the remaining tests are related to stuff I've touched (and the one failure isn't my fault).

Anything more to add, @vargaz @safern @ViktorHofer @akoeplinger ?

akoeplinger commented on eng/pipelines/mono/templates/build-job.yml:
```diff
@@ -101,6 +112,13 @@ jobs:
+        displayName: Download AOT offset files
+        inputs:
+          artifact: Mono_Offsets_Android
+          path: '$(Build.SourcesDirectory)/artifacts/obj/mono/offsetfiles'
```
do we still need this here if it's already in global-build-job.yml?

akoeplinger commented on eng/pipelines/runtime-official.yml:
```diff
@@ -122,6 +122,37 @@ stages:
+      - OSX_x64
+      - Linux_x64
+      jobParameters:
+        buildArgs: -s mono+libs+host+packs -c $(_BuildConfig)
```
do we really need the `libs` subset here?

akoeplinger commented on eng/pipelines/runtime-official.yml:
```diff
@@ -122,6 +122,37 @@ stages:
+      jobTemplate: /eng/pipelines/mono/templates/generate-offsets.yml
+      buildConfig: release
+      platforms:
+      - Android_x64
```
maybe add a comment here that this will actually generate offsets for all platforms, not just x64.

akoeplinger commented on eng/pipelines/runtime.yml:
```diff
@@ -183,6 +183,22 @@ jobs:
+    jobTemplate: /eng/pipelines/mono/templates/generate-offsets.yml
+    buildConfig: release
+    platforms:
+    - Android_x64
```
same comment here

akoeplinger commented on eng/pipelines/runtime.yml:
```diff
@@ -459,6 +475,34 @@ jobs:
 
+#
+# Build Mono release Android AOT cross-compiler
+# Only when mono changed
```
```suggestion
# Only when mono or installer changed
```

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -378,8 +380,20 @@
+
     <PropertyGroup>
-      <MonoLibClang Condition="'$(MonoLibClang)' == ''">$(XcodeDir)/Toolchains/XcodeDefault.xctoolchain/usr/lib/libclang.dylib</MonoLibClang>
+      <MonoLibClang Condition="$([MSBuild]::IsOSPlatform('OSX')) and '$(MonoLibClang)' == ''">$(XcodeDir)/Toolchains/XcodeDefault.xctoolchain/usr/lib/libclang.dylib</MonoLibClang>
```
we should probably use the libclang.dylib from the ANDROID_NDK_ROOT here too instead of the one from xcode.

akoeplinger reviewed (approved):
Looks great, left a couple comments but nothing that should hold off merging this :)
